### PR TITLE
RFC: Vulkan Object to hold parsed XML

### PR DESCRIPTION
**Pull request by spencer-lunarg**
Wrote up in the documentation, but the goal of this PR is to get feedback on the design before going further

The main idea is

1. `base_generator.py` is the only file that will need to understand the `reg.py`/`generator.py` file from the Vulkan-Headers
2. Anyone who will want to add generation code can just look at `vulkan_object.py` and grab what they want in a C-struct style format (using python dataclasses)

### CODE REVIEW

**Comment by juan-lunarg** [comment]
I think I'm on board with the idea proposed here. The main thing I'd ask is to keep in mind we plan on moving code generation to the utility libraries.

https://github.com/KhronosGroup/Vulkan-Utility-Libraries/issues/13

**Comment by charles-lunarg** [comment]
What I would like to see is more details on how the many 'things' (structs, struct members, enums, bitmasks, constants, function pointers, etc) are described and what data is available about them.

I have done a non-trivial amount of work parsing vk.xml into my own python data structures that makes it easy to iterate over them [here](https://github.com/cdgiessen/vk-module/blob/master/generate.py). The context is me trying to create my own C++ bindings ala vulkan-hpp. Didn't go that well, but thats besides the point.

I spent a lot of time in the command param/struct member stuff, because I wanted to extract a lot of details about the types. Like, is it const, is it a pointer to an array of string, a fixed length array, a dynamic array with length based on another parameter/member, so on and so forth. 

The other thing I spent a lot of time on was dependency tracking. "X struct needs Y type, which comes from either Z extension or version W, in which case use either the extension or the base type depending."

A comment about usability - while having C style structs is 100% fine, I think of the spec like a relational database. You have a table per 'type class' like enum, struct, union, function, then each row of the table is each instance of the 'type class', which contains keys (like the name), regular data (like the C type or an enum's value), and foreign keys to other rows, (like extension/version dependencies, dependent types, AKA any relationship with other parts of the spec).

So I'm not asking for an SQL query language to use when generating code, but I do treat codegen like that. This is very evident in my vulkaninfo autogen work, where I actively cull out any types that aren't necessary. In that case, I just skip generation of types not needed, but to figure that out I have to recursively go through the struct list until I have the set of all necessary structs. The usecase for vulkaninfo is a bit weird, since I purposefully only generate core, KHR, and EXT stuff.

But the point stands that when doing codegen, there will be times where the logic has a similar vibe to "select all types where any member contains a pointer to an array".

**Comment by spencer-lunarg** [comment]
> The main thing I'd ask is to keep in mind we plan on moving some code generation to the utility libraries.

Maybe the right move is to make the base generator and Vulkan Object part of the `registry` then as an optional opt-in

>  I think of the spec like a relational database

:100: 

The ultimate goal is to form the data into some sort of "lookup" instead of everyone having to recreate their own internal "lookup" database

File: scripts/generators/vulkan_object.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    name: str
+    alias: str
+    externsync: str
+    optional: str
+    selector: str
+    noautovalidity: str
+    type: str
+    def __init__(self, param):
+        self.name = param.find('name').text
+        self.type = textIfFind(param, 'type')
```
Comment (ncesario-lunarg) [inline scripts/generators/vulkan_object.py:107]: Will this also capture the `*` (or other "dangling  markup") for pointer types?

**Review by ncesario-lunarg — COMMENTED** [review]
This would be a welcome addition, but I guess my main comment/question is with respect to whether or not we can generate the data classes from reigstry.rnc, or if perhaps we could work towards that such that.

File: docs/generated_code.md [hunk 2]
```diff
@@ -36,12 +34,58 @@ NOTE: `VVL_CODEGEN` is `OFF` by default to allow users to build `VVL` via `add_s
...
+7. repeat step 3-6
+8. `SomethingOutputGenerator::endFile()`
+9. `OutputGenerator::endFile()`
+
+The big difference is `SomethingOutputGenerator` (ex. `CommandValidationOutputGenerator`) only has to be called  **once** at the end.
+This means each generator script doesn't have to worry about understanding how the `registry` file works.
+
+This is possible because of the new `class VulkanObject()`
+
+The `VulkanObject` makes use of the [Python 3.7 Dataclasses](https://docs.python.org/3/library/dataclasses.html) to enforce a schema so developers don't have to go guessing what each random object from the various `registry` is. Most of the schema is derived from the [Spec's registry.rnc](https://github.com/KhronosGroup/Vulkan-Docs/blob/main/xml/registry.rnc) file. This file is used to enforce the `vk.xml` schema and serves as a good understanding of what will be in each element of the XML.
```
[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply docs/generated_code.md:86]: So the issue is the `rnc` is just a schema for the XML

I don't think getting a 1-for-1 mapping from the XML is what we want, that is basically something Python's `ElementTree` could do

As Charles pointed out, the XML is basically a Database with many tables that need to interact on each other

I think maybe the issue is that **currently** I need use the `.rnc` to understand random `ElementTree` element handles I get... I think the goal is to move away from having to care at all about the XML and just focus on seeing what is in the `VulkanObject` class

**Comment by spencer-lunarg** [comment]
closing - will have more incremental PR in future to show

**Pull request closed without merging**